### ISSUE TITLE
Acc question about Espnet2.0

### ISSUE THREAD

**Issue by xvolica** [espnet/espnet#2439]
Recipe: hkust
Task: ASR

**Acc Question**
I am using hkust to compare Espnet2 and Espnet1, using same parameters, the espnet2 get worse result than espnet1, here is the result:

Ver | Corr | Sub | Del | Ins | Err | S.Err | tag
-- | -- | -- | -- | -- | -- | -- | --
Espnet1.0 | 78.9 | 17.4 | 3.7 | 3.5 | 24.5 | 73 | original
Espnet2.0 | 74.9 | 20.1 | 5 | 3.2 | 28.3 | 77.2 | lr=1.0,noam(25000)
Espnet2.0 | 76.5 | 18.8 | 4.7 | 3.3 | 26.8 | 75.9 | lr=0.005,warmup(25000)

And I try to adjust the lr, warmup-steps and so on, but it doesn't help. Is there any suggestion?

**Basic environments:**
 - OS information: Ununtu16.04 Linux 4.4.0-131-generic
 - python version: 3.8.5
 - espnet 0.9.1
 - pytorch 1.4.0
 - GPU V100 x8(CUDA 10.2)

**Comment by kamo-naoyuki** [comment]
About DP, 
It's interesting. I quickly tested DP, but I couldn't observe.
There are very small different parts between DP and DDP, so I don't have any idea except for Pytorch bug.
Could you test it with pytorch1.6.0?



About ACC,

HKUST is not good  to compare the result. This corpus is too large for benchmark.

- Could you show the raw sclite output? I mean, could you write also # Snt    # Wrd?
- Is it single GPU mode? If not, please compare with single GPU to avoid small difference between multi GPUs behaviour.
- What about batch-size? Did you adjust the fold_length to the same value?
- The model_size(= the attention dimension) is a constant parameter in the noam scheduler in espnet2, so you need to change it by yourlself and also change the parameter of warmup scheduler accordingly.

**Comment by kamo-naoyuki** [comment]
Sorry, I remembered DP doesn't work with pytorch1.6.0 or 1.5 because of espnet1's bug.

**Comment by xvolica** [comment]
I just solved the DP things, a small bug caused it that I only gave optimizer the encoder network.
As for acc, I'll try your suggestions, thx!

Here is a worse result in espnet2:
accum_grad: 2
lr: 0.00039528470752104737
warmup_steps: 25000
batch_size: 128
scheduler: warmuplr

|SPKR|#Snt|#Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|Sum/Avg|3999|3776|17.9|81.5|0.6|0.9|83.0|78.4|
|Mean|181.8|171.6|17.7|81.7|0.6|1.0|83.2|78.6|
|S.D.|27.8|27.3|6.4|6.2|1.0|1.5|6.8|6.8|
|Median|182.5|167.0|16.5|82.6|0.0|0.5|84.1|77.5|

**Comment by kamo-naoyuki** [comment]
What's the evalset? It has quite different number of sentence and words.

**Comment by xvolica** [comment]
Esp1-dev-cer
|SPKR|#Snt|#Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|  ----  | ----  |----  |----  |----  |----  |----  |----  |----  |
|Sum/Avg|5413|56154|78.7|17.8|3.6|3.0|24.3|69.5|
|Mean|112.8|1169.9|79.1|17.4|3.5|2.9|23.8|68.7|
|S.D.|26.9|473.9|7.9|6.8|1.6|2.4|9.5|13.7|
|Median|112.5|1181.0|80.8|16.1|3.2|2.3|21.4|69.0|

Esp2-dev-cer
|SPKR|#Snt|#Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|  ----  | ----  |----  |----  |----  |----  |----  |----  |----  |
|Sum/Avg|5413|56154|75.1|20.5|4.5|2.7|27.6|73.2|
|Mean|225.5|2339.8|75.2|20.4|4.4|2.7|27.4|73.4|
|S.D.|41.7|445.7|7.3|5.9|1.8|1.9|8.4|10.4|
|Median|220.5|2459.0|75.3|20.5|4.1|2.2|27.2|71.4|

**Comment by kamo-naoyuki** [comment]
LM? There are too large difference. Could you show the acc/loss plot?

**Comment by kamo-naoyuki** [comment]
To make clear, could you also show the each result with lm_weight=0?

**Comment by kamo-naoyuki** [comment]
And, could you attach your configuration of training?

**Comment by xvolica** [comment]
### train_asr_transformer.yaml
```
# network architecture
# encoder related
encoder: transformer
encoder_conf:
    input_layer: conv2d
    num_blocks: 12
    linear_units: 2048
    dropout_rate: 0.1
    output_size: 256
    attention_heads: 4
    attention_dropout_rate: 0.0

# decoder related
decoder: transformer
decoder_conf:
    input_layer: embed
    num_blocks: 6
    linear_units: 2048
    dropout_rate: 0.1

# hybrid CTC/attention
model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1
    length_normalized_loss: false

# optimization related
optim: adam
accum_grad: 2
grad_clip: 5
patience: 3
max_epoch: 50
optim_conf:
    lr: 1.0
scheduler: noamlr
scheduler_conf:
    model_size: 256
    warmup_steps: 25000

# minibatch related
batch_type: folded
batch_size: 512

# criterion
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 10
```
### acc/loss plot
The acc and loss look similar,both in trends and results, I can't upload images, so I put the final scores here.
```
Espnet1{
        "main/loss_ctc": 7.885093072652817,
        "main/loss_att": 4.848338074684143,
        "main/acc": 0.9178624668667399,
        "main/loss": 5.759364606142044,
        "validation/main/loss_ctc": 19.86078919063915,
        "validation/main/loss_att": 14.013803352009166,
        "validation/main/acc": 0.7863065673673411,
        "validation/main/loss": 15.767899383198131,
        "epoch": 19,
        "iteration": 10600,
}
Espnet2 20epoch results:
[train] loss=4.657, loss_att=4.077, loss_ctc=6.011, acc=0.928, lr_0=6.362e-04
[valid] loss=13.734, loss_att=12.140, loss_ctc=17.453, acc=0.784, cer=0.329, wer=0.790, cer_ctc=0.369
```

**Comment by kamo-naoyuki** [comment]
What about LM?

**Comment by kamo-naoyuki** [comment]
You forgot setting xavier_uniform initialization.
https://github.com/espnet/espnet/blob/d8e1bf011c4d791a5898b72f0f3f15eab3c95450/egs2/wsj/asr1/conf/tuning/train_asr_transformer2.yaml#L7

**Comment by xvolica** [comment]
This is the conf of LM, maybe batch_size is too large
```
optim: sgd
patience: 3
max_epoch: 20
batch_type: folded
batch_size: 1024 # 300 for word LMs
lm: seq_rnn
lm_conf:
    rnn_type: lstm
    nlayers: 2  # 1 for word LMs
    unit: 650   # 1000 for word LMs

best_model_criterion:
-   - valid
    - loss
    - min
keep_nbest_models: 1
```

**Comment by kamo-naoyuki** [comment]
it's better to change the lr because the default is different from espnet1. More nlayers may got better performance.

**Comment by xvolica** [comment]
I have done many experiments these days, and I still can not reach the best score in Espnet1.
- I change the batch_size to 64 in train_lm which get the best result.
|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|lm2/train_dev|3999|47130|77.6|17.8|4.6|3.3|25.7|74.6|
|lm2/dev|5413|56154|77.6|18.0|4.4|2.6|25.0|70.4|

- I changed the lr and nlayers then, but the result came down.
- I changed the lr, batch_size, fold_length, noam, accum_grad all same to the Espnet1.0, and using init: xavier_uniform, but the result is not very well.
|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|lm2/train_dev|3999|47130|76.6|18.5|4.9|3.0|26.4|75.1|
|lm2/dev|5413|56154|77.0|18.4|4.6|2.6|25.5|70.9|

**Comment by kamo-naoyuki** [comment]
The result seems enough. The relative difference (CER 25.0 vs 24.3) is small to accept and such difference can come from , even, random seed tuning. 
Even if all hyper parameters correspond to espnet1, we can't avoid all difference, i.e. randomness of initialization.

However, I feel we can improve LM part a little more in this case. 
If you'll tune this task more, the other idea is using the other optimizer or using the other initialization method, more small batchsize.
